### PR TITLE
Adds a connection pool for maintaining connections.

### LINKED ISSUE #51: Too many open files
Hi,

I've hit a "too many open files" error while publishing to the InfluxDB. I've noticed that publisher creates a new socket to the database for each publish action, but sockets are not being closed. I'm using Snap v0.13.0-beta.

```
root@node-8:~# lsof -p 56250 |grep 8086 |wc -l
527
```

snapd.conf:
```
---
log_level: 1
log_path: /var/log/snap
control:
  auto_discover_path: /etc/snap/plugin
  plugin_trust_level: 0
```

task file:
```
---
version: 1
schedule:
  type: 'simple'
  interval: '1s'
workflow:
  collect:
    metrics:
      /intel/linux/iostat/avg-cpu/%idle: {}
      /intel/linux/iostat/avg-cpu/%iowait: {}
      /intel/linux/iostat/avg-cpu/%nice: {}
      /intel/linux/iostat/avg-cpu/%steal: {}
      /intel/linux/iostat/avg-cpu/%system: {}
      /intel/linux/iostat/avg-cpu/%user: {}
      /intel/linux/iostat/device/ALL/%util: {}
      /intel/linux/iostat/device/ALL/avgqu-sz: {}
      /intel/linux/iostat/device/ALL/avgrq-sz: {}
      /intel/linux/iostat/device/ALL/await: {}
      /intel/linux/iostat/device/ALL/r_await: {}
      /intel/linux/iostat/device/ALL/r_per_sec: {}
      /intel/linux/iostat/device/ALL/rkB_per_sec: {}
      /intel/linux/iostat/device/ALL/rrqm_per_sec: {}
      /intel/linux/iostat/device/ALL/svctm: {}
      /intel/linux/iostat/device/ALL/w_await: {}
      /intel/linux/iostat/device/ALL/w_per_sec: {}
      /intel/linux/iostat/device/ALL/wkB_per_sec: {}
      /intel/linux/iostat/device/ALL/wrqm_per_sec: {}
      /intel/procfs/meminfo/MemUsed: {}
      /intel/procfs/meminfo/MemUsed_perc: {}
      /intel/procfs/meminfo/SwapFree: {}
      /intel/procfs/meminfo/SwapFree_perc: {}
      /intel/procfs/meminfo/MemFree: {}
      /intel/procfs/meminfo/MemFree_perc: {}
      /intel/procfs/load/min1: {}
      /intel/procfs/load/min5: {}
      /intel/procfs/load/min15: {}
      /intel/procfs/iface/br-mgmt/bytes_recv: {}
      /intel/procfs/iface/br-mgmt/bytes_sent: {}
      /intel/procfs/iface/br-storage/bytes_recv: {}
      /intel/procfs/iface/br-storage/bytes_sent: {}
      /intel/procfs/swap/all/cached_bytes: {}
      /intel/procfs/swap/all/cached_percent: {}
      /intel/procfs/swap/all/free_bytes: {}
      /intel/procfs/swap/all/free_percent: {}
      /intel/procfs/swap/all/used_bytes: {}
      /intel/procfs/swap/all/used_percent: {}
    process:
      -
        plugin_name: 'passthru'
        process: null
        publish:
          -
            plugin_name: 'influx'
            config:
              host: 
              port: 8086
              database: snap
              user: 
              password: 
```

Related log entries:
```
time="2016-04-28T12:22:45Z" level=debug msg="Batch submission of process and publish nodes" _block=work-jobs _module=scheduler-workflow count-process-nodes=0 count-publish-nodes=1 parent-node-type=processor task-id=f36c8347-4b7d-42e9-85da-3da2099b9062 task-name=Task-f36c8347-4b7d-42e9-85da-3da2099b9062 
time="2016-04-28T12:22:45Z" level=debug msg="Submitting publish job" _block=submit-publish-job _module=scheduler-workflow parent-node-type=processor publish-name=influx publish-version=-1 task-id=f36c8347-4b7d-42e9-85da-3da2099b9062 task-name=Task-f36c8347-4b7d-42e9-85da-3da2099b9062 
time="2016-04-28T12:22:45Z" level=debug msg="starting publisher job" _module=scheduler-job block=run content-type=snap.gob job-type=publisher plugin-config=map[password:{} port:{8086} user:{} database:{snap} host:{}] plugin-name=influx plugin-version=-1 
time="2016-04-28T12:22:45Z" level=debug msg="plugin selected" _module=control-routing block=select hitcount=1014 index="publisher:influx:v12:id1" pool size=1 strategy=least-recently-used 
time="2016-04-28T12:22:45Z" level=error msg="error with publisher job" _module=scheduler-job block=run content-type=snap.gob error="Publish call error: Post http://:8086/write?consistency=&db=snap&precision=s&rp=default: dial tcp :8086: socket: too many open files" job-type=publisher plugin-config=map[password:{} port:{8086} user:{} database:{snap} host:{}] plugin-name=influx plugin-version=-1 
time="2016-04-28T12:22:45Z" level=warning msg="Publish job failed" _block=submit-publish-job _module=scheduler-workflow parent-node-type=processor publish-name=influx publish-version=-1 task-id=f36c8347-4b7d-42e9-85da-3da2099b9062 task-name=Task-f36c8347-4b7d-42e9-85da-3da2099b9062 
```

### PR DESCRIPTION
**Pull request by lynxbat**
Fixes #51 - Adds a connection pool for maintaining connections to InfluxDB. Connections are unique by user, db, and URL. A watching function closing idle connections (30min).

### CODE REVIEW

File: influx/influx.go [hunk 1]
```diff
@@ -43,6 +45,21 @@ const (
 	defaultTimestampPrecision = "s"
 )
 
+var (
+	// The maximum time a connection can sit around unused.
+	maxConnectionIdle = time.Minute * 30
+	// How feequently idle connections are checked
```
Comment (thomastaylor312) [inline influx/influx.go]: "Frequently" instead of "feequently"

File: influx/influx.go [hunk 2]
```diff
@@ -43,6 +45,21 @@ const (
 	defaultTimestampPrecision = "s"
 )
 
+var (
+	// The maximum time a connection can sit around unused.
+	maxConnectionIdle = time.Minute * 30
```
Comment (thomastaylor312) [inline influx/influx.go]: Should this value be configurable?

File: influx/influx.go [hunk 3]
```diff
@@ -89,6 +106,25 @@ func (f *influxPublisher) GetConfigPolicy() (*cpolicy.ConfigPolicy, error) {
...
+		for k, c := range connPool {
+
+			if time.Now().Sub(c.LastUsed) > maxConnectionIdle {
+				m.Lock()
+				// Close the connection
+				c.close()
+				// Remove from the pool
+				delete(connPool, k)
+				m.Unlock()
+				fmt.Println("Removed an idle connection")
```
Comment (thomastaylor312) [inline influx/influx.go]: Change to use `logger`

Comment (jcooklin) [inline influx/influx.go]: No, it ages out connections that haven't been used in `maxConnectionIdle`.

Comment (lynxbat) [inline influx/influx.go]: The print line is. I will remove. Logging needs to be revisited for a plugin like this. Probably a caching of the last logger in a package-level var so we can write to the logger when not in the Publish Function.

File: influx/influx.go (hunk 2 above)
Comment (lynxbat) [inline influx/influx.go]: I debated that. But given how little impact this is (it is just an optimization) I decided to leave it out for now.

Comment (thomastaylor312) [inline influx/influx.go]: 👍

**Comment by thomastaylor312** [comment]
LGTM. Merging.

**Pull request merged**